Parse command line options

sl4mmy opened the pull request:
Added a simple command line option parser.  Currently only recognizes two options:
* ```-q|--quiet``` - suppress intermediate output
* ```-p|--print-classpath``` - print all artifacts in classpath format (useful for setting environment/Makefile variables that you can pass to ```java -cp```)

sl4mmy commented:
@joewalnes This ```-p|--print-classpath``` option can simplify Makefile-based Java builds: add a target to run ShavenMaven for resolving dependencies as a prerequisite of everything else, and capture its output in a variable suitable for passing to ```java -cp```

sl4mmy commented:
Unfortunately, this conflicts with #9.  If you want to merge them one-at-a-time I'm happy to rebase either one on top of the other and fix all the merge conflicts.

danielbodart commented on src/com/googlecode/shavenmaven/Dependencies.java:
```diff
@@ -49,11 +49,11 @@ public boolean update(File directory) {
     }
 
     public static void main(String[] args) throws Exception {
-        if (args.length == 0 || args.length > 2) {
```
Not so sure about this change, was there a specific reason to go this way or just cleanness?!

sl4mmy replied:
This commit was just laying the groundwork for later splitting command line arguments between options & args.  I'm happy to squash this commit into that later one if you prefer a cleaner/shorter commit history.

danielbodart commented on src/com/googlecode/shavenmaven/Configuration.java:
```diff
@@ -0,0 +1,62 @@
+
+public class Configuration {
+
+    private static final PrintStream DEVNULL = new PrintStream(new NoopOutputStream());
```
Could you use Streams.nullPrintStream() please?

danielbodart commented on src/com/googlecode/shavenmaven/Resolver.java:
```diff
@@ -47,17 +49,17 @@ public Resolver(File directory) {
         this(directory, System.out);
     }
 
-    public boolean resolve(Artifact artifact) throws Exception {
```
Not sure about this, if you don't like the boolean what about an Enum or do you pass a custom message somewhere?

sl4mmy replied:
Sorry, I'm obviously not familiar with TotallyLazy: does it provide a maybe monad like Optional?  That's probably the best approach.

danielbodart commented:
One thing @npryce asked for but I never finished was support for a -test option (prints same stuff but down't download anything) so Make could work out the dependencies. Is that something that you guys would like as well? (As you mentioned Make)

danielbodart commented:
Okay I've add some inline comments, please have a look.

I think we should try and merge the recursive option first as it's simpler then this one on top

sl4mmy commented:
I rebased this on top of the recursive change, and incorporated your feedback.  I also ended up switching the resolver to return an Option<String> now.  It seems really nice to me, but I'm curious to hear your thoughts on such a big API change.  Was your concern before more around the ugly UNRESOLVED constant, or is your preference to stick with the boolean API?

danielbodart commented:
Yea the constant seems pretty suckie, what do you return when it is
resolved?

On Mon, 21 Mar 2016 4:06 pm Kent R. Spillner, <notifications@github.com>
wrote:

> I rebased this on top of the recursive change, and incorporated your
> feedback. I also ended up switching the resolver to return an Option now.
> It seems really nice to me, but I'm curious to hear your thoughts on such a
> big API change. Was your concern before more around the ugly UNRESOLVED
> constant, or is your preference to stick with the boolean API?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/bodar/shavenmaven/pull/10#issuecomment-199358023>
>

sl4mmy commented:
I'll reply re: the -test option to the separate issue

sl4mmy commented:
I return option(destination.getPath())

https://github.com/bodar/shavenmaven/pull/10/commits/92184f6b2c1981259b517a3d81a399871cd11a7a#diff-819666a726f38a8ea4ca640f2447d87aR74

danielbodart commented on src/com/googlecode/shavenmaven/Dependencies.java:
```diff
@@ -40,37 +43,43 @@ public static Dependencies load(File file, PrintStream out) {
                 map(delete()).realise();
         final Resolver resolver = new Resolver(directory, out);
-        return artifacts.filter(not(Artifact.functions.existsIn(directory))).mapConcurrently(resolve(resolver)).forAll(is(true));
+        return artifacts.filter(not(Artifact.functions.existsIn(directory))).mapConcurrently(resolve(resolver));
```
Looking at this code again, I'm not sure the way you have currently added class path support is actually going to work, nor do I think the method signature actually makes sense.

Currently where do you handle building the class path when no dependencies download? Changing 'resolve' to return a class path seems weird as resolve is only called when something has changed.

Generating a class path should really just be completely separate piece of code with no code change to the existing API. I'd just have a new Class that iterates over the Artifacts.

Sorry I didn't catch this right at the start.

danielbodart commented:
Okay I started a clean room implementation of this, and realised it seems like more code for something that is probably good enough just look at the filesystem:

If you don't care about sources files:

Java `recursiveFiles("lib").filter(hasSuffix("jar"))`
Make `find lib -type f -name '*.jar' | tr '\n' ':'`

If you want to filter sources (because JavaC will recompile them if they are on the class path!)

Java `recursiveFiles("lib").filter(hasSuffix("jar").filter(not(hasSuffix("sources.jar")))`
Make `find lib -type f -name '*.jar' -not -name '*sources.jar'  | tr '\n' ':'`


If you wanted to get more hard core and match against the dependencies file you could do this:

Java 6+ (you can use the built in wildcard support):

`find build -type f -name '*.dependencies' | sed 's/build\(.*\).dependencies/lib\1\/*/' | tr '\n' ':'`

 < Java 6

`find build -type f -name '*.dependencies' | sed 's/build\(.*\).dependencies/lib\1/' | xargs -I{} find {} -type f -not -name '*sources.jar' | tr '\n' ':'`

Okay this has a requirement on shell tools but if you are using Make you are probably already happy with that.

Unless you have a requirement to know the class path **before** downloading the files, (i.e. from the actual dependencies files) this all seems like overkill for something that can be done simpler else where.

What ya think?

sl4mmy commented:
hehe, actually, that's exactly what we're doing today.  I wanted to push that into ShavenMaven because I thought it would simplify some of that for us.  Also, making ShavenMaven the canonical source of truth for the definitive list of dependencies avoids problems with multiple versions of some libraries showing up (for example, we have a dependency on an internal library that changes frequently).  The alternative to that would be frequently clean builds that purge ```lib/``` but then that means we're constantly downloading the same things over & over.

sl4mmy commented:
All that being said, though, I'm happy to table this for now.  We're just spiking a lot of this stuff out now and haven't really started using ShavenMaven in anger yet, so all of this is speculation and could turn out to be YAGNI.  Do you want to leave this PR open and wait for me to check back in in a couple weeks to see what our actual experience is like?

danielbodart commented:
SM will guarantee the lib folder matches the dependencies at all time (well
as long as you run it)

On Thu, 24 Mar 2016 at 15:38 Kent R. Spillner <notifications@github.com>
wrote:

> hehe, actually, that's exactly what we're doing today. I wanted to push
> that into ShavenMaven because I thought it would simplify some of that for
> us. Also, making ShavenMaven the canonical source of truth for the
> definitive list of dependencies avoids problems with multiple versions of
> some libraries showing up (for example, we have a dependency on an internal
> library that changes frequently). The alternative to that would be
> frequently clean builds that purge lib/ but then that means we're
> constantly downloading the same things over & over.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/bodar/shavenmaven/pull/10#issuecomment-200890817>
>

danielbodart commented:
Perfect, that sounds sensible

On Thu, 24 Mar 2016 at 15:40 Kent R. Spillner <notifications@github.com>
wrote:

> All that being said, though, I'm happy to table this for now. We're just
> spiking a lot of this stuff out now and haven't really started using
> ShavenMaven in anger yet, so all of this is speculation and could turn out
> to be YAGNI. Do you want to leave this PR open and wait for me to check
> back in in a couple weeks to see what our actual experience is like?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/bodar/shavenmaven/pull/10#issuecomment-200891770>
>
